### PR TITLE
Non-record: Internal control port on the PR180 stack

### PR DESCRIPTION
**Pull request by small-cactus**
## Summary

Non-record submission porting our late-attention internal-control method onto the PR180 stack.

This is **not** a record claim yet. The record run is still in development. This PR documents the port, the current training/eval path, and the runtime fixes needed to make the approach runnable on multi-GPU pods.

## Key ideas

1. **Late-attention internal control** on top of the PR180 recipe rather than replacing the existing stack.
2. **Detached gate-controlled attention scaling** with `gate_fuse_mode=k`.
3. **EMA-zscore energy normalization** with clipping to stabilize the control signal.
4. **Compile-path fixes** so the control recurrence does not blow up graph size in the public script.
5. **Fast iteration support** for public-track debugging: capped validation for smoke runs and eval progress logging.

## Files

- `records/track_10min_16mb/2026-03-20_10L_Int5MLP_MuonWD04_SWA50/train_gpt.py`
  - Internal-control port
  - Compile isolation fixes for the recurrent control path
  - Multi-GPU grad accumulation override
  - Fastcheck validation cap and eval progress logging

## Reproducibility

Fastcheck example:

```bash
CONTROL_ENABLED=1 \
CONTROL_DETACH_GATES=1 \
CONTROL_GATE_FUSE_MODE=k \
CONTROL_ENERGY_NORM=ema_zscore \
CONTROL_ENERGY_CLIP=0.99 \
CONTROL_APPLY_TO=attn \
CONTROL_LATE_LAYERS=3 \
CONTROL_WARMUP_STEPS=8 \
CONTROL_SCAN_BLOCK_SIZE=16 \
VAL_LOSS_EVERY=0 \
VAL_MAX_TOKENS=4194304 \
COMPILE_MODEL=0 \
COMPILE_ZERPOWER=1 \
COMPILE_CONTROL_SCAN=0 \
torchrun --standalone --nproc_per_node=5 train_gpt.py
```

## Status

- Public-track port is running on multi-GPU hardware.
- Fastcheck runs are working after fixing the control compile path and the silent full-validation stall.
- Full record-size evaluation is still in development.
- No record claim in this PR.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: Internal control port on the PR180 stack

**BPB:** (not parsed — see PR title) | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `3cf231ffdde2`, file `records/track_10min_16mb/2026-03-20_10L_Int5MLP_MuonWD04_SWA50/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 3.69s, dim=512, layers=10, vocab=1024, code=70825 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 3.69s, dim=512, layers=10, vocab=1024, code=70825 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
